torch/_tensor.py

```diff
@@ -1318,6 +1318,97 @@ def to_sparse_coo(self):
         """
         return self.to_sparse()
 
+    def dim_order(self):
+        """
+
+        dim_order() -> tuple
+
+        Returns a tuple of int describing the dim order or physical layout of :attr:`self`.
+
+        Args:
+            None
+
+        Dim order represents how dimensions are laid out in memory,
+        starting from the outermost to the innermost dimension.
+
+        Thus, the conversion from strides is done by sorting the strides
+        from larger to smaller since the dimension with the largest stride
+        is the outermost and the dimension with the smallest stride is the innermost.
+
+        For example, tensor with sizes = (3, 5, 2) and strides = (5, 1, 15), implies
+        physical layout of (2, 0, 1). Dimension order of (2, 0, 1) can be obtained
+        by sorting strides from large to smaller.
+
+        Example::
+            >>> torch.empty((2, 3, 5, 7)).dim_order()
+            (0, 1, 2, 3)
+            >>> torch.empty((2, 3, 5, 7), memory_format=torch.channels_last).dim_order()
+            (0, 2, 3, 1)
+
+        .. warning::
+            The dim_order tensor API is experimental and subject to change.
+
+        """
+
+        # NB:
+        # - Based on the implementation in TensorIterator.cpp
+        # - Should have similar behavior (esp in ambiguous cases) to,
+        #   torch._prims_common.compute_elementwise_output_logical_to_physical_perm
+
+        if self.layout != torch.strided:
+            raise RuntimeError("dim_order is only supported for strided tensors.")
+
+        ndim = self.ndim
+
+        if ndim == 0:
+            return tuple()
+
+        if ndim == 1:
+            return (0,)
+
+        if self.is_contiguous(memory_format=torch.contiguous_format):
+            return tuple(range(ndim))
+
+        shape = self.shape
+
+        def should_swap(idx_a, idx_b):
+            stride_a = self.stride()[idx_a]
+            stride_b = self.stride()[idx_b]
+
+            if stride_a == 0 or stride_b == 0:
+                return 0
+
+            if stride_a < stride_b:
+                return -1
+
+            if stride_a > stride_b:
+                return 1
+
+            # stride_a == stride_b
+            if shape[idx_a] > shape[idx_b]:
+                return 1
+
+            # Note: this case is hit when all strides are non-zero and equal, and all
+            # dimensions have the same length
+            return 0
```

Comment:
Nit: these can be combined in a more structured way:
```suggestion
            if stride_a < stride_b:
                return -1
            elif stride_a > stride_b:
                return 1
            # stride_a == stride_b
            elif shape[idx_a] > shape[idx_b]:
                return 1
            else:
            # Note: this case is hit when all strides are non-zero and equal, and all
            # dimensions have the same length
                return 0
```